lighthouse-core/report/html/renderer/details-renderer.js

```diff
@@ -39,6 +39,13 @@ class DetailsRenderer {
     this._templateContext = context;
   }
 
+  /**
+   * @param {LH.Artifacts.FullPageScreenshot} fullPageScreenshot
+   */
+  setFullPageScreenshot(fullPageScreenshot) {
```

Comment:
`setTemplateContext` was added as a last minute architecture concession to handle the fact that we couldn't always use `document` to clone templates (because we were going into DevTools) and we weren't prepared to refactor so close to 2.0. It's not really a good pattern and it's not self obvious how to use these or even that they're there to be used in the first place.

I don't have a lot of ideas off hand (besides adding some documentation), but any reason to not add it to the constructor, at least, so at least the suggestion is right there in your face whenever creating a `DetailsRenderer`? (I'd have to look into to it to see if it's still true, but I believe `setTemplateContext` was separated specifically because we didn't necessarily have the context yet at construction time)

Reply:
It'd have to be an optional parameter, otherwise it's technically a breaking change. Ideally the constructor parameter would never be optional (it should be `FullPageScreenshot|null`). If in the future we have more progressive enhancement data for the report, the constructor would become a lengthy list of optional parameters, which I think is a poor API.

I think a `setter` fn like this communicates well that it's an optional enhancement, is a better API if scaled, and avoids breaking changes.

Reply:
> It'd have to be an optional parameter, otherwise it's technically a breaking change. Ideally the constructor parameter would never be optional (it should be `FullPageScreenshot|null`).

since`full-page-screenshot` comes through as an audit it'll to always have to be optional, because there's no guarantee the audit will have been included in the run.

> I think a setter fn like this communicates well that it's an optional enhancement, is a better API if scaled, and avoids breaking changes.

it seems like the intended use is, if a screenshot exists, to *always* set it up like
```js
const detailsRenderer = new DetailsRenderer(this._dom);
detailsRenderer.setFullPageScreenshot(fullPageScreenshot);
```
and then to never call `setFullPageScreenshot()` again (e.g. if it was, the background image url wouldn't work for the first or the second set of screenshots, depending on when everything is set up).

That's basically acting like a constructor parameter over two calls. The `options` object is a very popular JS idiom for customization of a function call/class construction, and that's usually where users of the class would look first for what options are available (as opposed to its various methods).

I'm not saying it's a great API :) but I think it's the best of the options we've come up with so far.

Reply:
options sgtm